plugin/src/main/java/org/opensearch/sql/plugin/datasource/OpenSearchDataSourceMetadataStorage.java

```diff
@@ -0,0 +1,244 @@
+      = "datasources-index-settings.yml";
+  private static final Logger LOG = LogManager.getLogger();
+  private final Client client;
+  private final ClusterService clusterService;
+
+  private final Encryptor encryptor;
+
+  /**
+   * This class implements DataSourceMetadataStorage interface
+   * using OpenSearch as underlying storage.
+   *
+   * @param client         opensearch NodeClient.
+   * @param clusterService ClusterService.
+   * @param encryptor      Encryptor.
+   */
+  public OpenSearchDataSourceMetadataStorage(Client client, ClusterService clusterService,
+                                             Encryptor encryptor) {
+    this.client = client;
+    this.clusterService = clusterService;
+    this.encryptor = encryptor;
+  }
+
+  @Override
+  public List<DataSourceMetadata> getDataSourceMetadata() {
+    if (!this.clusterService.state().routingTable().hasIndex(DATASOURCE_INDEX_NAME)) {
+      createDataSourcesIndex();
+      return Collections.emptyList();
+    }
+    return searchInDataSourcesIndex(QueryBuilders.matchAllQuery());
+  }
+
+  @Override
+  public Optional<DataSourceMetadata> getDataSourceMetadata(String datasourceName) {
+    if (!this.clusterService.state().routingTable().hasIndex(DATASOURCE_INDEX_NAME)) {
+      createDataSourcesIndex();
+    }
+    return searchInDataSourcesIndex(QueryBuilders.termQuery("name", datasourceName))
+        .stream()
+        .findFirst()
+        .map(x -> this.encryptDecryptAuthenticationData(x, false));
+  }
+
+  @Override
+  public void createDataSourceMetadata(DataSourceMetadata dataSourceMetadata) {
+    encryptDecryptAuthenticationData(dataSourceMetadata, true);
+    if (!this.clusterService.state().routingTable().hasIndex(DATASOURCE_INDEX_NAME)) {
+      createDataSourcesIndex();
+    }
+    IndexRequest indexRequest = new IndexRequest(DATASOURCE_INDEX_NAME);
+    indexRequest.id(dataSourceMetadata.getName());
+    ActionFuture<IndexResponse> indexResponseActionFuture;
+    try (ThreadContext.StoredContext storedContext = client.threadPool().getThreadContext()
+        .stashContext()) {
+      indexRequest.source(XContentParserUtils.convertToXContent(dataSourceMetadata));
+      indexResponseActionFuture = client.index(indexRequest);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+    IndexResponse indexResponse = indexResponseActionFuture.actionGet();
+    if (indexResponse.getResult().equals(DocWriteResponse.Result.CREATED)) {
+      LOG.debug("DatasourceMetadata : {}  successfully created", dataSourceMetadata.getName());
+    }
+  }
+
+  @Override
+  public void updateDataSourceMetadata(DataSourceMetadata dataSourceMetadata) {
+    throw new UnsupportedOperationException("will be supported in future.");
+  }
+
+  @Override
+  public void deleteDataSourceMetadata(String datasourceName) {
+    throw new UnsupportedOperationException("will be supported in future.");
+  }
+
+  private void createDataSourcesIndex() {
```

Comment:
does user need `create index` permission? or it is coverd by `cluster:admin/opensearch/datasources/`?

Reply:
Line 134
  ```
  try (ThreadContext.StoredContext storedContext = client.threadPool().getThreadContext()
          .stashContext()) {
        createIndexResponseActionFuture = client.admin().indices().create(createIndexRequest);
      }
```
     We stash existing context, so client treats this code being invoked by plugin rather than user.

Reply:
Create an issue for query execution plan.